Writerside/topics/raptor.md

```diff
@@ -1,49 +1,61 @@
-routing and planning in large transportation networks.
-
-## Flowchart
+The **RAPTOR** (Round-based Public Transit Routing) algorithm is a powerful algorithm to compute optimal routes in
+public transportation systems, especially for large and complex networks. Its primary objective is to efficiently
+determine the best route options based on factors like arrival time and number of transfers. RAPTOR is particularly
+useful for real-time journey planning due to its speed and adaptability.
+
+### Key Concepts of RAPTOR:
+
+- **Initialization**: The algorithm starts at the origin station and time specified by the user. Initially, it only
+  considers the starting station, marking it for the first round of exploration.
+
+- **Round-based Search**: The search process is broken into several rounds. In each round, the algorithm evaluates all
+  transport lines connected to the current list of stations. It computes the earliest possible arrival times at the
+  stations along each line and adds them to the list of stations for the next round.
+
+- **Updating Arrival Times**: As the algorithm processes new stations, it updates their earliest arrival times. If it
+  finds a route to a station that is faster than the previously known route, the arrival time and corresponding route
+  are updated.
+
+- **Termination Condition**: The algorithm stops when it reaches the destination station with the optimal arrival time
+  or when no further improvements can be made after a set number of rounds.
+
+- **Tracing the Optimal Route**: Once the algorithm concludes, the optimal route is reconstructed by tracing the updated
+  arrival times back from the destination to the start station.
+
+RAPTOR is highly efficient because it leverages the structure of public transit systems, where it is often faster to
+remain on the same line rather than switching between multiple lines. This characteristic makes RAPTOR particularly
+suitable for large-scale transportation networks where real-time routing is essential.
+
+## Sequence Diagram RAPTOR
 
 ```mermaid
-graph TD
-    A[Start at departure stop] --> B{Round-based search}
-    B -->|Found new stop| C[Update earliest arrival time]
-    C --> D{Check for termination}
-    D -->|Not terminated| B
-    D -->|Terminated| E[Backtrack route]
-    E --> F[End]
+sequenceDiagram
+    participant User
+    participant Algorithm
+    participant Stops
+    participant Routes
+    participant Timetable
+    User ->> Algorithm: Request journey (departure stop, destination stop, departure time)
+    Algorithm ->> Stops: Initialize list of stops to visit and visited stops
+    Algorithm ->> Stops: Set departure stop and time
+
+    loop Round-Based Search
+        Algorithm ->> Routes: Get transit lines from current stops
+        Routes ->> Timetable: Calculate earliest arrival time for each stop
+        Timetable ->> Algorithm: Return arrival times for unvisited stops
```

Comment:
I think this return can be implicitly derived and therefore does not need to be added.

Reply:
I added two reworked and simpler sequence diagrams. Which do you prefer ? @clukas1 @munterfi 

A)
```mermaid
sequenceDiagram
  participant User
  participant RAPTOR Algorithm
  participant Routes

  User->>RAPTOR Algorithm: Request journey plan
  RAPTOR Algorithm->>Routes: Initialize routes and transfers
  Routes-->>RAPTOR Algorithm: Routes initialized

  loop For each round k
    RAPTOR Algorithm->>Routes: Process marked stops and routes
    Routes-->>RAPTOR Algorithm: Return earliest trip information
    RAPTOR Algorithm->>RAPTOR Algorithm: Update arrival times
  end

  RAPTOR Algorithm->>User: Return Pareto-optimal journeys

```

B)
```mermaid
sequenceDiagram
  participant User
  participant RAPTOR Algorithm
  participant Routes
  participant Pareto Set

  User->>RAPTOR Algorithm: Instantiate RAPTOR Algorithm (using schedule)
  User->>RAPTOR Algorithm: Request Journey Plan (source, target, dep. time)

  RAPTOR Algorithm->>Routes: Sort Routes by Departure Time
  Routes-->>RAPTOR Algorithm: Sorted Routes

  loop Per Round (k = 1, 2, ...)
    RAPTOR Algorithm->>Routes: Scan Connections in Order
    Routes-->>RAPTOR Algorithm: Update Arrival Times
    RAPTOR Algorithm->>RAPTOR Algorithm: Relax Footpaths and Transfers
    RAPTOR Algorithm->>Pareto Set: Check for Pareto Dominance
    Pareto Set-->>RAPTOR Algorithm: Update Pareto Optimal Journeys
  end

  RAPTOR Algorithm->>User: Return Final Pareto Optimal Journeys
```

Reply:
I'm really sorry, but I have issues with both...

A): 

- Initialize Routes and Transfers does not happen after the user requests connections, so it doesn't really fit in this sequence as the user has no role in this process -> Raptor is built with the application.
- The Routes object is probably a suboptimal name, as you use it for transfers, stop times, stops, routes. Maybe RaptorData?
- The return arrays for both "Routes initialized" and "Return earliest trip information" seem unnecessary.
- Connections are not scanned, routes are scanned. But routes are not really scanned on the routes class/component, but in the construct we called RouteScanner or in pure algorithm reference is a step that the algorithm performs. So theoretically an arrow to itself with Scan Routes.

B)
- Again RAPTOR is not instantiated by the user.
- Routes are not sorted by departure times, trips are, but also this happens during instantiation and not when the user requests a connection.
- Connections are not scanned in order, routes are scanned while traversing trips in departure time ascending order.
- Same as above I think the Routes container should be replaced or even removed, as it's not really Routes we are only interested in.
- I would again also make the arrow for scanRoutes show on the RAPTOR algorithm itself because this is essentially part of the Algorithm (we've subdivided this in classes FootpahRelaxer, RouteScanner etc but it's all the algorithm).
- Footpaths and transfers are the same (as long as we don't include our own creation of same stop transfers), therefore I would rename the arrow to "Relax Footpaths".
- Regarding footpaths, it might be good to include that footpaths are also relaxed in round 0 (outside the loop), before route scanning.
- There is nothing I can think of that would resemble a pareto set in our algorithm, the way we decide that we have all pareto optimal solutions is only by the marked stops at the end of the round.

Reply:
```mermaid
graph TD
    A[Input: Source stop ps, Target stop pt, Departure time t] --> B[Initialization]
    B --> C1[Set all stop times to infinity]
    B --> C2[Set source stop time to departure time]
    B --> C3[Mark source stop]

    C3 --> D[For each round k]
    D --> E[Clear queue Q]

    E --> F[For each marked stop p]
    F --> G[For each route r serving stop p]

    G --> H1{Is route r already in Q?}
    H1 --> I1[Yes: Replace stop if current stop comes earlier in route]
    H1 --> I2[No: Add route and stop to Q]
    I1 --> J[Unmark stop p]
    I2 --> J[Unmark stop p]

    J --> K[For each route r in Q]
    K --> L[Set trip t to none]

    L --> M[For each stop pi on route r starting from p]
    M --> N1{Can the stop time be improved?}
    N1 --> O1[Yes: Update stop time and mark stop pi]
    N1 --> O2[No: Continue]

    O1 --> P1[Can we catch an earlier trip at stop pi?]
    O2 --> P1

    P1 --> Q1{Is previous stop time less than trip departure time?}
    Q1 --> R1[Yes: Set trip t to earlier trip]
    Q1 --> R2[No: Continue]

    R1 --> S[Check footpaths]

    S --> T[For each marked stop p]
    T --> U[For each footpath from p to another stop]
    U --> V[Update time of destination stop if faster]
    V --> W[Mark the destination stop]

    W --> X[Check stopping criterion]
    X --> Y1{Are no stops marked?}
    Y1 --> Z1[Yes: Stop]
    Y1 --> Z2[No: Continue to next round]
```

Reply:
```mermaid
graph TD
    A[Input: Source stop ps, Target stop pt, Departure time t] --> B[Initialization]
    B --> C[Mark source stop]
    
    C --> D[Main Loop: For each round k]
    D --> E[Accumulate routes serving marked stops]
    E --> F[For each route, traverse stops]
    F --> G[Update stop times if improvement found]
    G --> H[Check for footpaths between stops]
    H --> I[Mark newly reachable stops]
    
    I --> J{Are no stops marked?}
    J --> K[Yes: Stop]
    J --> L[No: Continue to next round]

```